File: frontend/mobile/lib/services/subscription_service.dart
import 'package:dio/dio.dart';

import '../core/api_client.dart';
import '../models/api_error.dart';
import '../models/responses/paged_result.dart';
import '../models/responses/subscription_response.dart';
import 'api_exception.dart';

/// The buyer's own recurring reservations (monthly parking). Mirrors `frontend/web`'s
/// SubscriptionService.
class SubscriptionService {
  bool _isSuccess(int? code) => code != null && code >= 200 && code < 300;

  Never _handleError(Response response) {
    ApiError apiError;
    try {
      final body = response.data;
      apiError = body is Map<String, dynamic>
          ? ApiError.fromJson(body)
          : ApiError(message: body?.toString());
    } catch (_) {
      apiError = ApiError(message: response.data?.toString());
    }
    throw ApiException(statusCode: response.statusCode ?? 0, apiError: apiError);
  }

  /// GET /api/subscriptions/mine
  Future<PagedResult<SubscriptionResponse>> getMine({int page = 0, int pageSize = 20}) async {
    final response = await apiClient.get(
      'Subscriptions/mine',
      queryParameters: {'page': page, 'pageSize': pageSize},
    );
    if (!_isSuccess(response.statusCode)) _handleError(response);
    return PagedResult.fromJson(response.data as Map<String, dynamic>, SubscriptionResponse.fromJson);
  }

  /// POST /api/subscriptions/{id}/cancel — schedules cancellation at the end of the paid-for
  /// period. The subscription stays active until the provider confirms it has actually ended, so
  /// the buyer keeps the month they already paid for.
  Future<void> cancel(String id) async {
    final response = await apiClient.post('Subscriptions/$id/cancel');
    if (!_isSuccess(response.statusCode)) _handleError(response);
  }
}

File: frontend/web/src/app/core/services/subscription.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable, map } from 'rxjs';

import { environment } from '../../../environments/environment';
import { PagedResult } from '../models/catalog.models';
import { Subscription, toSubscriptionStatus } from '../models/subscription.models';

@Injectable({ providedIn: 'root' })
export class SubscriptionService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = environment.apiBaseUrl;

  /** GET /api/subscriptions/mine — the caller's own recurring reservations. */
  getMine(page = 0, pageSize = 10): Observable<PagedResult<Subscription>> {
    return this.http
      .get<PagedResult<Subscription>>(`${this.baseUrl}/subscriptions/mine`, {
        params: { page: String(page), pageSize: String(pageSize) },
      })
      .pipe(
        map((result) => ({
          ...result,
          // status is an ordinal on the wire, same as every other enum — left un-normalized it
          // renders as a bare "0" in the status chip.
          items: result.items.map((s) => ({ ...s, status: toSubscriptionStatus(s.status) })),
        })),
      );
  }

  /**
   * POST /api/subscriptions/{id}/cancel — schedules cancellation at the end of the paid-for period.
   * The subscription stays Active until the provider confirms it has actually ended, so the buyer
   * keeps the month they already paid for.
   */
  cancel(id: string): Observable<void> {
    return this.http.post<void>(`${this.baseUrl}/subscriptions/${id}/cancel`, {});
  }
}
